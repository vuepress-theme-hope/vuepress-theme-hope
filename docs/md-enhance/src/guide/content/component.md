---
title: Component
icon: puzzle-piece
---

You can add components easily in Markdown content.

<!-- more -->

## Settings

::: code-tabs#language

@tab TS

```ts {8} title=".vuepress/config.ts"
import { mdEnhancePlugin } from "vuepress-plugin-md-enhance";

export default {
  plugins: [
    mdEnhancePlugin({
      // Enable component support
      component: true,
    }),
  ],
};
```

@tab JS

```js {8} title=".vuepress/config.js"
import { mdEnhancePlugin } from "vuepress-plugin-md-enhance";

export default {
  plugins: [
    mdEnhancePlugin({
      // Enable component support
      component: true,
    }),
  ],
};
```

:::

<!-- #region after -->

## Usage

You can use component fence block to add a component into your markdown content. Both YAML and JSON format props data are supported:

- YAML <Badge text="Recommended" type="tip" />:

  ````md
  ```component ComponentName
  # component data here
  ```
  ````

- JSON:

  ````md
  ```component ComponentName
  {
    // component data here
  }
  ```
  ````

## Demo

::: md-demo VPCard

```component VPCard
title: Mr.Hope
desc: Where there is light, there is hope
logo: https://mister-hope.com/logo.svg
link: https://mister-hope.com
background: rgba(253, 230, 138, 0.15)
```

```component VPCard
{
  "title": "Mr.Hope",
  "desc": "Where there is light, there is hope",
  "logo": "https://mister-hope.com/logo.svg",
  "link": "https://mister-hope.com",
  "background": "rgba(253, 230, 138, 0.15)"
}
```

:::

`<VPCard>` here is a global component.

The above code blocks are equivalent to:

```md
<VPCard
  title="Mr.Hope"
  desc="Where there is light, there is hope"
  logo="https://mister-hope.com/logo.svg"
  link="https://mister-hope.com"
  background="rgba(253, 230, 138, 0.15)"
/>
```

<!-- #endregion after -->
