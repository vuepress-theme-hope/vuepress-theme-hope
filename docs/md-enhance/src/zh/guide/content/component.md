---
title: 组件
icon: puzzle-piece
---

你可以在 Markdown 中轻松添加组件。

<!-- more -->

## 配置

::: code-tabs#language

@tab TS

```ts {8} title=".vuepress/config.ts"
import { mdEnhancePlugin } from "vuepress-plugin-md-enhance";

export default {
  plugins: [
    mdEnhancePlugin({
      // 开启组件支持
      component: true,
    }),
  ],
};
```

@tab JS

```js {8} title=".vuepress/config.js"
import { mdEnhancePlugin } from "vuepress-plugin-md-enhance";

export default {
  plugins: [
    mdEnhancePlugin({
      // 开启组件支持
      component: true,
    }),
  ],
};
```

:::

<!-- #region after -->

## 使用

你可以使用 component 代码块来在 Markdown 中添加组件。YAML 和 JSON 的数据格式均受支持:

- YAML <Badge text="推荐" type="tip" />:

  ````md
  ```component 组件名称
  # 组件数据
  ```
  ````

- JSON:

  ````md
  ```component 组件名称
  {
    // 组件数据
  }
  ```
  ````

## 案例

::: md-demo 快速使用 VPCard 组件

```component VPCard
title: Mr.Hope
desc: Where there is light, there is hope
logo: https://mister-hope.com/logo.svg
link: https://mister-hope.com
background: rgba(253, 230, 138, 0.15)
```

```component VPCard
{
  "title": "Mr.Hope",
  "desc": "Where there is light, there is hope",
  "logo": "https://mister-hope.com/logo.svg",
  "link": "https://mister-hope.com",
  "background": "rgba(253, 230, 138, 0.15)"
}
```

:::

这里的 `<VPCard>` 是已经全局注册的卡片组件。

上方的代码块和下方等价:

```md
<VPCard
  title="Mr.Hope"
  desc="Where there is light, there is hope"
  logo="https://mister-hope.com/logo.svg"
  link="https://mister-hope.com"
  background="rgba(253, 230, 138, 0.15)"
/>
```

<!-- #endregion after -->
