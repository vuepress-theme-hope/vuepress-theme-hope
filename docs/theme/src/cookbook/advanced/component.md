---
title: Using Components
icon: puzzle-piece
order: 1
category:
  - Tutorial
  - Customize
tag:
  - Components
  - Customize
---

This tutorial will guide you on how to use Vue components and Vue syntax in your VuePress project.

<!-- more -->

## Import Vue Components Globally

### Register via `@vuepress/plugin-register-components`

You can automatically register components via the `@vuepress/plugin-register-components` plugin.

For details about how to use the plugin, see [Official Documentation](https://v2.vuepress.vuejs.org/reference/plugin/register-components.html).

### Register via ClientConfigFile

You can create `.vuepress/client.ts` and register components manually.

```ts
// .vuepress/client.ts
import { defineClientConfig } from "@vuepress/client";
import MyComponent from "./MyComponent.vue";

export default defineClientConfig({
  enhance: ({ app, router, siteData }) => {
    app.component("MyComponent", MyComponent);
  },
});
```

## Using Vue syntax and components in Markdown

You can use Vue syntax directly in Markdown, for details see [VuePress â†’ Markdown](../vuepress/markdown.md#using-vue-in-markdown).

If you need to import Vue components in Markdown, please note that you cannot use relative paths to import, nor write multiple `<script>` blocks.

::: info Markdown with Vue SFC

Every Markdown file is first compiled to HTML and then converted to a Vue Single File Component (SFC). In other words, you can write Markdown files like Vue SFC:

`<script>` and `<style>` tags are directly treated as tags in Vue SFC. In other words, they are promoted from the `<template>` tag to the top level of the SFC.
All content other than `<script>` and `<style>` tags will be compiled to HTML and then treated as `<template>` tags in Vue SFC.

Since Vue single-file components can only contain one `<script>` tag, you should avoid using more than one `<script>` tag in VuePress Markdown.

In addition, since Markdown will be converted to Vue single-file components in the cache directory, any relative path import will be invalid in Vue SFC.

:::

In order to import your own components correctly, you need to create aliases for them, you can do this with the `alias` option:

```ts
// .vuepress/config.ts
import { getDirname, path } from "@vuepress/utils";

const __dirname = getDirname(import.meta.url);

export default {
  alias: {
    "@MyComponent": path.resolve(__dirname, "components/MyComponent.vue"),
  },
};
```

```md
<MyComponent />

<script setup lang="ts">
import MyComponent from "@MyComponent";
</script>
```

It's a bit more complicated, but if the component is only used in one page, importing like this has advantages:

- Global import means that the component code needs to be loaded when VuePress is initialized, that is, when visiting the first page
- Importing in Markdown causes the component code to be included in the page code so that it will only be loaded when visiting the page
