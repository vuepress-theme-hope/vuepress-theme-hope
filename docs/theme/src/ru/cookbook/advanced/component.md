---
title: Использование компонентов
icon: puzzle-piece
order: 1
category:
  - Руководство
  - Кастомизация
tag:
  - Компоненты
  - Кастомизация
---

В этом руководстве вы узнаете, как использовать компоненты Vue и синтаксис Vue в вашем проекте VuePress.

<!-- more -->

## Импорт компонентов Vue глобально

### Регистрация через `@vuepress/plugin-register-components`

Вы можете автоматически регистрировать компоненты с помощью плагина `@vuepress/plugin-register-components`.

Подробнее об использовании плагина смотрите в [Официальной документации](https://v2.vuepress.vuejs.org/reference/plugin/register-components.html).

### Регистрация через ClientConfigFile

Вы можете создать `.vuepress/client.ts` и зарегистрировать компоненты вручную.

```ts
// .vuepress/client.ts
import { defineClientConfig } from "@vuepress/client";
import MyComponent from "./MyComponent.vue";

export default defineClientConfig({
  enhance: ({ app, router, siteData }) => {
    app.component("MyComponent", MyComponent);
  },
});
```

## Использование синтаксиса и компонентов Vue в Markdown

Вы можете использовать синтаксис Vue непосредственно в Markdown. Для конкретного использования смотрите [VuePress → Markdown](../vuepress/markdown.md#Use -vue in -markdown-)

Если вам нужно импортировать компоненты Vue в Markdown, обратите внимание, что вы не можете использовать относительные пути для импорта или писать несколько блоков `<script>`.

::: info Markdown с Vue SFC

Каждый файл Markdown сначала компилируется в HTML, а затем преобразуется в компонент одного файла Vue (SFC). Другими словами, вы можете писать файлы Markdown, такие как Vue SFC:

Теги `<script>` и `<style>` напрямую обрабатываются как теги в Vue SFC. Другими словами, они продвигаются из тега `<template>` на верхний уровень SFC.
Весь контент, кроме тегов `<script>` и `<style>`, будет скомпилирован в HTML, а затем обработан как теги `<template>` в Vue SFC.

Поскольку однофайловые компоненты Vue могут содержать только один тег `<script>`, вам следует избегать использования более одного тега `<script>` в VuePress Markdown.

Кроме того, поскольку Markdown будет преобразован в однофайловые компоненты Vue в каталоге кеша, любой импорт относительного пути будет недопустимым в Vue SFC.

:::

Чтобы корректно импортировать собственные компоненты, вам необходимо создать для них псевдонимы, сделать это можно с помощью опции `alias`:

```ts
// .vuepress/config.ts
import { getDirname, path } from "@vuepress/utils";

const __dirname = getDirname(import.meta.url);

export default {
  alias: {
    "@MyComponent": path.resolve(__dirname, "components/MyComponent.vue"),
  },
};
```

```md
<MyComponent />

<script setup lang="ts">
import MyComponent from "@MyComponent";
</script>
```

Это немного сложнее, но если компонент используется только на одной странице, такой импорт имеет преимущества:

- Глобальный импорт означает, что код компонента нужно загрузить при инициализации VuePress, то есть при посещении первой страницы
- Импорт в Markdown приводит к включению кода компонента в код страницы, чтобы он загружался только при посещении страницы
