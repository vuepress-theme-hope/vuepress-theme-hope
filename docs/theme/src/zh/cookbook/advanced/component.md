---
title: 使用组件
icon: puzzle-piece
order: 1
category:
  - 教程知识
  - 自定义
tag:
  - 自定义
  - 组件
---

本教程将指引你如何在 VuePress 项目中使用 Vue 组件与 Vue 语法。

<!-- more -->

## 全局导入 Vue 组件

### 通过 `@vuepress/plugin-register-components` 注册组件

你可以通过 `@vuepress/plugin-register-components` 插件来自动注册组件。

插件的使用方法详见 [官方文档](https://v2.vuepress.vuejs.org/zh/reference/plugin/register-components.html)。

### 通过 ClientConfigFile 注册

你可以通过创建 `.vuepress/client.ts` 手动注册组件。

```ts
// .vuepress/client.ts
import { defineClientConfig } from "@vuepress/client";
import MyComponent from "./MyComponent.vue";

export default defineClientConfig({
  enhance: ({ app, router, siteData }) => {
    app.component("MyComponent", MyComponent);
  },
});
```

## 在 Markdown 中使用 Vue 语法与组件

你可以直接在 Markdown 中使用 Vue 语法。关于具体的使用方式，详见 [VuePress → Markdown](../vuepress/markdown.md#在-markdown-中使用-vue)

如果你需要在 Markdown 中导入 Vue 组件，请注意你不能使用相对路径进行导入，也不能书写多个 `<script>` 块。

::: info Markdown 与 Vue SFC

每一个 Markdown 文件，首先都会编译为 HTML ，然后转换为一个 Vue 单文件组件 (SFC) 。换句话说，你可以像写 Vue SFC 一样来写 Markdown 文件:

`<script>` 和 `<style>` 标签会直接被当作 Vue SFC 中的标签。换句话说，它们是从 `<template>` 标签中提升到了 SFC 的顶层。
所有 `<script>` 和 `<style>` 标签的以外的内容，会先被编译为 HTML ，然后被当作 Vue SFC 的 `<template>` 标签。

由于 Vue 单文件组件只能包含一个 `<script>` 标签，你应该避免在 VuePress Markdown 中使用多于一个 `<script>` 标签。

另外由于 Markdown 会被转换为缓存目录下的 Vue 单文件组件，任何相对路径的导入会在 Vue SFC 中失效。

:::

为了正确导入自己的组件，你需要为它们创建别名，你可以通过 `alias` 选项实现这一点:

```ts
// .vuepress/config.ts
import { getDirname, path } from "@vuepress/utils";

const __dirname = getDirname(import.meta.url);

export default {
  alias: {
    "@MyComponent": path.resolve(__dirname, "components/MyComponent.vue"),
  },
};
```

```md
<MyComponent />

<script setup lang="ts">
import MyComponent from "@MyComponent";
</script>
```

这固然复杂一些，但是如果你所使用的组件仅在这个页面使用，这样导入更具有优势:

- 全局导入意味着组件代码需要在 VuePress 初始化，也就是访问首个页面时载入
- 在 Markdown 中导入会使得组件代码包含在页面代码中，这样它仅会在访问这个页面时被导入
