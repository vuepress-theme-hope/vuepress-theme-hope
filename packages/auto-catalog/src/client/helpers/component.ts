import type { App, Component } from "vue";
import { inject } from "vue";

declare const __VUEPRESS_DEV__: boolean;

export type AutoCatalogIconComponent = Component<{ icon: string }>;

let autoCatalogIconComponent: Component = () => null;

const autoCatalogSymbol = Symbol(__VUEPRESS_DEV__ ? "auto-catalog" : "");

export const defineAutoCatalogIconComponent = (
  options: AutoCatalogIconComponent
): void => {
  autoCatalogIconComponent = options;
};

export const useAutoCatalogIconComponent = (): AutoCatalogIconComponent =>
  inject(autoCatalogSymbol)!;

export const injectAutoCatalogIconComponent = (app: App): void => {
  app.provide(autoCatalogSymbol, autoCatalogIconComponent);
};
