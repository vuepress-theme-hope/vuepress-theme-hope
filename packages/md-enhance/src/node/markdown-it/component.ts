import { load } from "js-yaml";
import type { Options, PluginSimple, Token } from "markdown-it";
import type { RenderRule } from "markdown-it/lib/renderer.js";
import type { MarkdownEnv } from "vuepress/markdown";

import { stringifyProp } from "./utils.js";
import { logger } from "../utils.js";

const getComponentRender =
  (name: string): RenderRule =>
  (
    tokens: Token[],
    index: number,
    _options: Options,
    { filePathRelative }: MarkdownEnv,
  ): string => {
    const token = tokens[index];
    const { content } = token;

    let config: unknown = null;

    if (content.trim().startsWith("{"))
      try {
        config = <unknown>JSON.parse(content);
      } catch (err) {
        // Do nothing
      }
    else
      try {
        config = load(content);
      } catch (err) {
        // Do nothing
      }

    if (config) return `<${name} v-bind='${stringifyProp(config)}' />`;

    logger.error(
      `Invalid ${name} config${
        filePathRelative ? ` found in ${filePathRelative}` : ""
      }:
${content}
`,
    );

    return "";
  };

export const component: PluginSimple = (md) => {
  // Handle ```component  blocks
  const { fence } = md.renderer.rules;

  md.renderer.rules.fence = (...args): string => {
    // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
    const [tokens, index] = args;
    const { info } = tokens[index];

    const [fenceName, componentName] = info.split(" ", 2);

    if (fenceName === "component") {
      const render = getComponentRender(componentName);

      return render(...args);
    }

    return fence!(...args);
  };
};
