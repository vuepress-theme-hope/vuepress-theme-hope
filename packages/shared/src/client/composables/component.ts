import type { App } from "vue";
import { camelize, capitalize, getCurrentInstance } from "vue";

export const hasGlobalComponent = (name: string, app?: App): boolean => {
  const globalComponents = (app?._instance || getCurrentInstance())?.appContext
    .components;

  if (!globalComponents) return false;

  return (
    name in globalComponents ||
    camelize(name) in globalComponents ||
    capitalize(camelize(name)) in globalComponents
  );
};
