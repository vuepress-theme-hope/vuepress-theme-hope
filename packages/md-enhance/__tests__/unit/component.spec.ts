import MarkdownIt from "markdown-it";
import { describe, expect, it } from "vitest";

import { component } from "../../src/node/markdown-it/component.js";

describe("component", () => {
  const markdownIt = MarkdownIt({ linkify: true }).use(component);

  it("Should resolve component fence", () => {
    const result1 = markdownIt.render(
      `
\`\`\`component VPCard
title: A card
desc: A card desc
logo: https://example.com/logo.png
link: https://example.com
color: "#000"
\`\`\`
`,
      {},
    );

    const result2 = markdownIt.render(
      `
\`\`\`component VPCard
{
  "title": "A card",
  "desc": "A card desc",
  "logo": "https://example.com/logo.png",
  "link": "https://example.com",
  "color": "#000"
}
\`\`\`
`,
      {},
    );

    expect(result1).toContain("VPCard");
    expect(result1).toMatchSnapshot();
    expect(result2).toContain("VPCard");
    expect(result2).toMatchSnapshot();
  });

  it("Should not throw with invalid syntax", () => {
    const result1 = markdownIt.render(
      `
\`\`\`component VPCard
title: a
title: b
\`\`\`
`,
      {},
    );

    const result2 = markdownIt.render(
      `
\`\`\`component VPCard
title: a
title: b
\`\`\`
`,
      {},
    );

    expect(result1).toEqual("");
    expect(result2).toEqual("");
  });

  it("Should drop when receiving a invalid syntax", () => {
    const result = markdownIt.render(
      `
\`\`\`component VPCard
{a:1}
\`\`\`
`,
      {},
    );

    expect(result).toMatch("");
  });

  it("Should not break markdown fence", () => {
    const result = markdownIt.render(
      `
\`\`\`js
const a = 1;
\`\`\`
`,
      {},
    );

    expect(result).toMatch(/<pre.*>[\s\S]*<\/pre>/);
    expect(result).toMatchSnapshot();
  });
});
