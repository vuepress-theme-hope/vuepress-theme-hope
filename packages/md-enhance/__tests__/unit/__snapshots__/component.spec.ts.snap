// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`component > Should not break markdown fence 1`] = `
"<pre><code class="language-js">const a = 1;
</code></pre>
"
`;

exports[`component > Should resolve component fence 1`] = `"<VPCard v-bind='{"title":"A card","desc":"A card desc","logo":"https://example.com/logo.png","link":"https://example.com","color":"#000"}' />"`;

exports[`component > Should resolve component fence 2`] = `"<VPCard v-bind='{"title":"A card","desc":"A card desc","logo":"https://example.com/logo.png","link":"https://example.com","color":"#000"}' />"`;
